lib/rules/no-useless-return.js

```diff
@@ -274,6 +296,22 @@ module.exports = {
                 }
                 scopeInfo.uselessReturns.push(node);
             },
+            "BlockStatement:exit"(node) {
+                if (
+                    !isTryStatement(node.parent) ||
+                    !scopeInfo.upper ||
+                    !scopeInfo.uselessReturns
+                ) {
+                    return;
+                }
+
+                const lastUseselessReturn = scopeInfo.uselessReturns[scopeInfo.uselessReturns.length - 1];
+
+
+                if (lastUseselessReturn && isLastNode(lastUseselessReturn, node, sourceCode)) {
+                    scopeInfo.upper.uselessReturns.push(lastUseselessReturn);
+                }
```

Comment:
Indeed, will go back to work and add that case to tests

Reply:
@mdjermanovic do you think is better to found another way of fixing it? I can think about using `TryStatement:exit` to check if the try statement is the last block within the function, otherwise remove the false positive - but this sound just like a workaround for the problem. 🤔

Reply:
The root cause of the problem seems to be that [code path analysis](https://eslint.org/docs/latest/developer-guide/code-path-analysis) considers segments containing `return`s from the try block as possible previous segments for segments in the catch block. This is logically incorrect, as the execution after `return;` cannot go into the catch block, but fixing this in the code path analysis would be a breaking change.

Also, even if we fix this in the code path analysis, there would still be false negatives with finally blocks.

```js
/* eslint no-useless-return: ["error"] */

function foo() {
  try {
    return; // false negative
  } finally {
    bar();
  }
}

```

In this case, code path analysis is logically correct because `bar();` is executed after the `return;`, but this `return;` is still useless.

I think we could try something like this in the `no-useless-return` rule:

1. Maintain a list of try blocks for which their catch/finally blocks are currently being traversed:
    * In `"TryStatement > BlockStatement.block:exit"`, push that block statement to an array in `scopeInfo`.
    * In `TryStatement:exit`, pop.
2. When [removing useless returns](https://github.com/eslint/eslint/blob/6bef1350e692c818c55c6d2074c12506e98cdf4f/lib/rules/no-useless-return.js#L169-L172), keep those that are in range of blocks from 1.

Reply:
@mdjermanovic can you open an issue describing the code path analysis bug? We should try to fix that in the next major release.

Reply:
I'll open an issue for code path analysis. Either way, try-catch-finally might require special-casing in this particular rule anyways, so I think it's also fine to try to fix https://github.com/eslint/eslint/issues/7481 directly in the rule in a minor release, as intended by this PR.